stan/math/opencl/kernel_generator/indexing.hpp

```diff
@@ -0,0 +1,282 @@
+#ifndef STAN_MATH_OPENCL_KERNEL_GENERATOR_indexing_HPP
+#define STAN_MATH_OPENCL_KERNEL_GENERATOR_indexing_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/opencl/kernel_generator/type_str.hpp>
+#include <stan/math/opencl/kernel_generator/name_generator.hpp>
+#include <stan/math/opencl/kernel_generator/operation_cl.hpp>
+#include <stan/math/opencl/kernel_generator/as_operation_cl.hpp>
+#include <stan/math/opencl/kernel_generator/is_kernel_expression.hpp>
+#include <string>
+#include <utility>
+
+namespace stan {
+namespace math {
+
+/** \addtogroup opencl_kernel_generator
+ *  @{
+ */
+
+/**
+ * Represents indexing of a matrix with two matrices of indices. `indexing(mat,
+ * col_index, row_index)[i,j] = mat[col_index[i,j], row_index[i,j]]`
+ * @tparam T_mat expression type of the matrix to index
+ * @tparam T_col_index expression type of the column index
+ * @tparam T_row_index expression type of the row index
```

Comment:
Order should match the code
```suggestion
 * @tparam T_row_index expression type of the row index
 * @tparam T_col_index expression type of the column index
```